File: Quorum3Plugins/TextPlugin/src/plugins/quorum/Libraries/Language/Types/Integer.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package plugins.quorum.Libraries.Language.Types;

import quorum.Libraries.Language.Object$Interface;
import quorum.Libraries.Language.Types.Integer$Interface;

/**
 * Integer plugin for handling Integer class functions for objects and primitives.
 * 
 * @author Melissa Stefik
 */
public class Integer {
    public java.lang.Object $me = null;
    private int integer = -1;
    
    public void SetValueNative(int value){
        integer = value;
    }
    
    public static boolean PrimitiveEquals(int self, Object$Interface obj) throws quorum.Libraries.Language.Errors.Error{
        Integer$Interface t = (Integer$Interface)obj;
        return self == t.GetValue();
    }
    
    public static double PrimitiveGetNumber(int self){
        return self;
    }
    
    public static String PrimitiveGetText(int self){
        return "" + self;
    }
    
    public static quorum.Libraries.Language.Support.CompareResult$Interface PrimitiveCompare(int self, quorum.Libraries.Language.Object$Interface obj){
        quorum.Libraries.Language.Support.CompareResult r = new quorum.Libraries.Language.Support.CompareResult();
        Integer$Interface t = (Integer$Interface)obj;
        if(self == t.GetValue()){
            r.result = r.EQUAL;
        } else if(self > t.GetValue()){
            r.result = r.LARGER;
        } else {
            r.result = r.SMALLER;
        }
        return r;
    }
    
    public int BitCount(){
        return java.lang.Integer.bitCount(integer);
    }
    
    public static int PrimitiveBitCount(int self){
        return java.lang.Integer.bitCount(self);
    }
    
    public int HighestOneBit(){
        return java.lang.Integer.highestOneBit(integer);
    }
            
    public static int PrimitiveHighestOneBit(int self){
        return java.lang.Integer.highestOneBit(self);
    }
    
    public int LowestOneBit(){
        return java.lang.Integer.lowestOneBit(integer);
    }
    
    public static int PrimitiveLowestOneBit(int self){
        return java.lang.Integer.lowestOneBit(self);
    }
    
    public int LeadingZeros(){
        return java.lang.Integer.numberOfLeadingZeros(integer);
    }
    
    public static int PrimitiveLeadingZeros(int self){
        return java.lang.Integer.numberOfLeadingZeros(self);
    }
    
    public int TrailingZeros(){
        return java.lang.Integer.numberOfTrailingZeros(integer);
    }
    
    public static int PrimitiveTrailingZeros(int self){
        return java.lang.Integer.numberOfTrailingZeros(self);
    }
    
    public int Reverse(){
        return java.lang.Integer.reverse(integer);
    }
    
    public static int PrimitiveReverse(int self){
        return java.lang.Integer.reverse(self);
    }
    
    public int RotateLeft(int value){
        return java.lang.Integer.rotateLeft(integer, value);
    }
    
    public static int PrimitiveRotateLeft(int self, int points){
        return java.lang.Integer.rotateLeft(self, points);
    }
    
    public int RotateRight(int value){
        return java.lang.Integer.rotateRight(integer, value);
    }
    
    public static int PrimitiveRotateRight(int self, int points){
        return java.lang.Integer.rotateRight(self, points);
    }
    
    public String GetBinary(){
        return java.lang.Integer.toBinaryString(integer);
    }
    
    public static String PrimitiveGetBinary(int self){
        return java.lang.Integer.toBinaryString(self);
    }
    
    public String GetHex(){
        return java.lang.Integer.toHexString(integer);
    }
    
    public static String PrimitiveGetHex(int self){
        return java.lang.Integer.toHexString(self);
    }
    
    public String GetOctal(){
        return java.lang.Integer.toOctalString(integer);
    }
    
    public static String PrimitiveGetOctal(int self){
        return java.lang.Integer.toOctalString(self);
    }
    
    public static int PrimitiveGetMaximumValue(int self){
        return java.lang.Integer.MAX_VALUE;
    }
    
    public static int PrimitiveGetMinimumValue(int self){
        return java.lang.Integer.MIN_VALUE;
    }
    
    public int GetHashCode() {
        return integer;
    }
    
    public int PrimitiveGetHashCode(int self) {
        return self;
    }
}

File: Quorum3/Library/Standard/Plugins/javascript/Libraries/Language/Types/Integer.js
function plugins_quorum_Libraries_Language_Types_Integer_(optional) {
    this.integer = 0;
    
    if(optional === undefined) {
    } else {
        this.integer = optional;
    }
    this.SetValueNative$quorum_integer = function (value) {
        this.integer = value;
    };
    
    this.BitCount = function() {
        var text = Number(this.integer).toString(2)
        var count = 0;
        for(i = 0; i < text.length; i++) {
            var char = text.charAt(i);
            if(char == "1") {
                count++;
            }
        }
        return count;
    };
    
    //system action GetHashCode() returns integer
    this.GetHashCode = function() {
        return this.integer.GetHashCode$IntegerPrimitive();
    };
    
    this.GetBinary = function() {
        return this.integer.GetBinary$IntegerPrimitive();
    };
    
    //system action BitCount returns integer
    this.BitCount = function() {
        return this.integer.BitCount$IntegerPrimitive();
    };
    //
    //system action HighestOneBit returns integer
    this.HighestOneBit = function() {
        return this.integer.HighestOneBit$IntegerPrimitive();
    };
    //system action LowestOneBit returns integer
    this.LowestOneBit = function() {
        return this.integer.LowestOneBit$IntegerPrimitive();
    };
    //system action LeadingZeros returns integer
    this.LeadingZeros = function() {
        return this.integer.LeadingZeros$IntegerPrimitive();
    };
    //system action TrailingZeros returns integer
    this.TrailingZeros = function() {
        return this.integer.TrailingZeros$IntegerPrimitive();
    };
    //system action Reverse returns integer
    this.Reverse = function() {
        return this.integer.Reverse$IntegerPrimitive();
    };
    //system action RotateLeft(integer value) returns integer
    this.RotateLeft = function() {
        return this.integer.RotateLeft$IntegerPrimitive();
    };
    //system action RotateRight(integer value) returns integer
    this.RotateRight = function() {
        return this.integer.RotateRight$IntegerPrimitive();
    };
    //system action GetBinary returns text
    this.GetBinary = function() {
        return this.integer.GetBinary$IntegerPrimitive();
    };
    //system action GetHex returns text
    this.GetHex = function() {
        return this.integer.GetHex$IntegerPrimitive();
    };
    //system action GetOctal returns text
    this.GetOctal = function() {
        return this.integer.GetOctal$IntegerPrimitive();
    };
}

Number.prototype.Equals$quorum_Libraries_Language_Object$IntegerPrimitive = function (value) {
    var b = global_CheckCast(value, "Libraries.Language.Types.Integer");

    var me = this.valueOf();
    var other = b.GetValue();
    if ((me == other)) {
        return true;
    } else {
        return false;
    }
};

//system action GetHashCode() returns integer
Number.prototype.GetHashCode$IntegerPrimitive = function () {
    return 0;
};
//system action BitCount returns integer
Number.prototype.BitCount$IntegerPrimitive = function() {
    var text = Number(this.valueOf()).toString(2);
    var count = 0;
    for(i = 0; i < text.length; i++) {
        var char = text.charAt(i);
        if(char == "1") {
            count++;
        }
    }
    return count;
};

//system action HighestOneBit returns integer
Number.prototype.HighestOneBit$IntegerPrimitive = function () {
    return 0;
};
//system action LowestOneBit returns integer
Number.prototype.LowestOneBit$IntegerPrimitive = function () {
    return 0;
};
//system action LeadingZeros returns integer
Number.prototype.LeadingZeros$IntegerPrimitive = function () {
    return 0;
};
//system action TrailingZeros returns integer
Number.prototype.TrailingZeros$IntegerPrimitive = function () {
    return 0;
};
//system action Reverse returns integer
Number.prototype.Reverse$IntegerPrimitive = function () {
    return 0;
};
//system action RotateLeft(integer value) returns integer
Number.prototype.RotateLeft$IntegerPrimitive = function (value) {
    return 0;
};
//system action RotateRight(integer value) returns integer
Number.prototype.RotateRight$IntegerPrimitive = function (value) {
    return 0;
};
//system action GetBinary returns text
Number.prototype.GetBinary$IntegerPrimitive = function() {
    return this.valueOf().toString(2);
};

//system action GetHex returns text
Number.prototype.GetHex$IntegerPrimitive = function () {
    return 0;
};
//system action GetOctal returns text
Number.prototype.GetOctal$IntegerPrimitive = function () {
    return 0;
};

Number.prototype.GetNumber$IntegerPrimitive = function() {
  return this.valueOf() * 1.0;
};

Number.prototype.GetText$IntegerPrimitive = function() {
  return this.valueOf() * "";
};

Number.prototype.GetMaximumValue$IntegerPrimitive = function() {
  return  2147483647;
};

Number.prototype.GetMinimumValue$IntegerPrimitive = function() {
  return  -2147483648;
};

Number.prototype.Compare$quorum_Libraries_Language_Object$IntegerPrimitive = function(value) {
    var result = new quorum_Libraries_Language_Support_CompareResult_();
    var b = global_CheckCast(value, "Libraries.Language.Types.Integer");
    
    var me = this.valueOf();
    var other = b.GetValue();
    if ((me == other)) {
        result.result = result.EQUAL;
    }
    else if ((me < other)) {
        result.result = result.SMALLER;
    }
    else {
        result.result = result.LARGER;
    }
    return result;
};

Number.prototype.GetValue$IntegerPrimitive = function() {
    return this.valueOf();
};
